### PR TITLE
Use std::chrono for the time to rotate destination of addr messages + tests

### PR DESCRIPTION
**Pull request by naumenkogs**
We currently assign a destination peer for relaying particular addresses of nodes every 24 hours, and then rotate. This is done for rate-limiting (ultimately for privacy leak reduction I think?).

Before this change, 24 hours was defined as uint. I replaced it with std::chrono, which is mockable and type-safe.

Also added couple tests for this behavior.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds rotate_interval = std::chrono::duration_cast<std::chrono::seconds>(ROTATE_ADDR_RELAY_DEST_INTERVAL);
```
Comment (sipa) [inline src/net_processing.cpp:1379]: Is this duration_cast needed? I think conversion should be automatic.

Comment (naumenkogs) [reply src/net_processing.cpp:1379]: You are right. I spent 30 minutes fighting with this code, and couldn't come up with something better.

But after you messaged I tried once more and it worked. Will force push in a second.

**Comment by naumenkogs** [comment]
@sipa btw, take a look at how hashAddr was fed to the hasher twice (with a shift, yeah).
I removed it, although I don't know why it was there in the first place... Feeding it once like I do should be sufficient.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
...
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    uint64_t time_seed = current_time.count() / rotate_interval.count();
```
Comment (MarcoFalke) [inline src/net_processing.cpp:1381]:
```suggestion
    const uint64_t time_seed {GetTime<std::chrono::seconds>().count() / std::chrono::seconds{ROTATE_ADDR_RELAY_DEST_INTERVAL}.count()};
```

Everything in one line with compile-time checks enabled would look like this, btw

Comment (naumenkogs) [reply src/net_processing.cpp]: can we keep it to 2 lines? I think what you suggest is hard to read.

**Comment by sipa** [comment]
@naumenkogs It was very inscrutable code, but the construction with double-hashing the address hash was by design.

The idea is that the address both influences the hash directly, but also determines when the "reset time of the day" is (by including hashaddr in `(time + hashaddr)/DAY`, not every address will change salt simultaneously throughout the day).

The shifting is pointless, we should get rid of it. It seems to be a silly evolution of this 2010 Satoshi code: 5cbf75324d1509a1262b65c5073314a4da3f6d77.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -1373,7 +1375,10 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    uint64_t time_seed = current_time / ROTATE_ADDR_RELAY_DEST_INTERVAL;
+    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr).Write(time_seed);
```
Comment (sipa) [inline src/net_processing.cpp:1381]:
```suggestion
    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr).Write((current_time + hashAddr) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```

Comment (naumenkogs) [reply src/net_processing.cpp:1381]: Yeah, I'm not 100% sure that this measure is effective against anything, but it won't hurt to have a less symmetry.
This mehaviour _may_ leak extra information if someone tries to analyze how addresses are flowing, but since we don't have any real evidence, we probably should stick to the existing reasoning.

Had to slightly adjust the code you suggested, because adding uint to chrono type doesn't work right away.

**Comment by naumenkogs** [comment]
Updated the code with suggestion by @sipa, and simplified tests.

They are a bit loosened now, but more robust to failures caused by randomness. 
Loosened in a sense they don't check the exact *daily* logic, because we can't predict when rotation happens for a given address (well, we can, but it's probably not worth to re-code all choose destination logic in python)

File: src/net_processing.cpp [hunk 4]
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = (current_time + std::chrono::hours(hashAddr)) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```
Comment (MarcoFalke) [inline src/net_processing.cpp:1381]: Previously `hashAddr` was of type seconds, now it is type hours. What was the reason to change it?

Comment (naumenkogs) [reply src/net_processing.cpp:1381]: I think i changed it while trying something else. There is no real difference, as long as it results is on order of hours or more. Otherwise, it won't add enough randomness.

Comment (MarcoFalke) [reply src/net_processing.cpp:1381]: You will get rid of the "reset time of the day" feature. Generally I don't like refactoring commits to also change behavior.

Comment (naumenkogs) [reply src/net_processing.cpp]: Okay, made it seconds again so that we don't change the behavior. Although I think we won't loose any properties if hours are used.

File: src/net_processing.cpp [hunk 5]
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = (current_time + std::chrono::seconds(hashAddr)) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```
Comment (MarcoFalke) [inline src/net_processing.cpp:1381]:
```suggestion
    uint64_t time_addr = (current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```

Maybe try this to avoid https://travis-ci.org/github/bitcoin/bitcoin/jobs/675780723#L4756

File: src/net_processing.cpp [hunk 6]
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = (current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```
Comment (MarcoFalke) [inline src/net_processing.cpp:1381]:
```suggestion
    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```


```
 node0 stderr net_processing.cpp:1381:26: runtime error: implicit conversion from type 'typename common_type<long, long>::type' (aka 'long') of value -28739051454386 (64-bit, signed) to type 'uint64_t' (aka 'unsigned long') changed the value to 18446715334658097230 (64-bit, unsigned)

File: test/functional/p2p_addr_relay.py [hunk 7]
```diff
@@ -48,24 +53,72 @@ def run_test(self):
...
                 'sending addr (301 bytes) peer=1',
         ]):
             addr_source.send_and_ping(msg)
-            self.nodes[0].setmocktime(int(time.time()) + 30 * 60)
+            mocked_time += 30 * 60
+            self.nodes[0].setmocktime(mocked_time)
             addr_receiver.sync_with_ping()
 
+        self.log.info('Check that within 24 hours an addr relay destination is rotated at most once')
+        mocked_time = int(time.time()) + 100000
```
Comment (jonatack) [inline test/functional/p2p_addr_relay.py:77]: I don't grok what/why regarding the 100000 value. It would be nice to comment and/or hoist it to a variable.

Same for 200000 at line 100.

File: test/functional/p2p_addr_relay.py [hunk 8]
```diff
@@ -48,24 +53,72 @@ def run_test(self):
...
+            with self.nodes[0].assert_debug_log([
+                'received: addr (31 bytes) peer=0',
+            ]):
+                addr_source.send_and_ping(msg)
+                mocked_time += 2 * 60 * 60
+                self.nodes[0].setmocktime(mocked_time)
+                for receiver in addr_receivers:
+                    receiver.sync_with_ping()
+        nodes_received_addr = [node for node in addr_receivers if node.received_addr]
+        assert(len(nodes_received_addr) <= 4)
```
[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/p2p_addr_relay.py:97]: Why the value of 4 here and line 118? Perhaps add a comment.

**Review by jonatack — COMMENTED** [review]
(Tested) Concept ACK.

Some questions and style/PEP8 comments below, feel free to ignore.

File: src/net_processing.cpp [hunk 9]
```diff
@@ -1433,7 +1435,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```
[the remark this answers is not in the document]
Comment (naumenkogs) [reply src/net_processing.cpp:1441]: We iterated over this line a couple times with @MarcoFalke, so I'm out of ideas here.
Open to suggestions!

**Comment by naumenkogs** [comment]
@jonatack addressed all the things.

**Review by jonatack — COMMENTED** [review]
ACK 72d7e24

Thanks for clarifying the tests! I spent some time messing around with them, refactored, and sped them up from 26 to 7 seconds on my machine. Feel free to use/steal/ignore: https://github.com/jonatack/bitcoin/commits/refactor-p2p_addr_relay-tests

File: src/net_processing.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (sipa) [reply src/net_processing.cpp:1441]: Annoyingly, I think this is UB, as hashAddr may exceed 2^63, and casting a larger number to a signed 64-bit integer would be outside its range (and overflowing signed integers is undefined behavior).

Suggestion: `uint64_t time_addr = (current_time + std::chrono::seconds{hashAddr % ROTATE_ADDR_RELAY_DEST_INTERVAL.count()}) / ROTATE_ADDR_RELAY_DEST_INTERVAL;`, and make `ROTATE_ADDR_RELAY_DEST_INTERVAL` to be of type `std::chrono::seconds`.

**Comment by sipa** [comment]
utACK 72d7e2400759dacfc8860341d5bb90459e363a81. I only casually reviewed the tests.

**Comment by MarcoFalke** [comment]
ACK 72d7e2400759dacfc8860341d5bb90459e363a81

**Comment by fanquake** [comment]
Aside from this needing rebase for conflicts, can you rebase on master so that new CIs run. That should remove appveyor as well.

**Comment by MarcoFalke** [comment]
@naumenkogs Is this up for grabs?

**Comment by naumenkogs** [comment]
Okay, it was hard to rebase @jonatack test contributed, but I just did it. Ready for review.

**Review by theStack — COMMENTED** [review]
Concept ACK

The first commits replaces `GetTime()` and the timing constants related to the randomization of ADDR messages with the mockable variants, i.e. it enables mocking but there shouldn't be any change in behaviour if mocking is not used. Will take a deeper look on the test soon, just left some nits (feel free to ignore).

**Comment by MarcoFalke** [comment]
@naumenkogs Is this up for grabs?

**Comment by naumenkogs** [comment]
@MarcoFalke I mean, it's up-to-date (apart from those little nits) and reviewable.
I probably don't have energy to nudge people for reviewing this, so if that makes it "up for grabs", then yes.

**Comment by MarcoFalke** [comment]
It would be good to reply to the comments. Either addressing them or denying/closing them.

**Comment by MarcoFalke** [comment]
Or alternatively submit the test in a separate pull request, since this is refactoring, so the test should already pass on master.

File: src/net_processing.cpp [hunk 10]
```diff
@@ -1700,7 +1702,11 @@ void PeerManagerImpl::RelayAddress(NodeId originator,
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = m_connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```
Comment (sdaftuar) [inline src/net_processing.cpp:1708]: Can you explain why it is safe / defined behavior to perform a `static_cast<uint64_t>(std::chrono::seconds)`?  I can't figure out from reading https://en.cppreference.com/w/cpp/chrono/duration whether this is supposed to be defined, but maybe I'm missing something.

Comment (sdaftuar) [reply src/net_processing.cpp:1708]: Sorry for the noise, @sipa explained this to me (the result of the division is being cast, which is not a duration but a number).

File: src/net_processing.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (sipa) [reply src/net_processing.cpp:1708]: This isn't resolved actually; it's still casting `hashAddr` from `uint64_t` to `int64_t`, which is implementation-defined behavior (I was wrong claiming earlier that it was UB).

[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:1708]: We're constructing a `std::chrono::seconds` from an `int64_t`, but according to https://en.cppreference.com/w/cpp/chrono/duration, `std::chrono::seconds` is only guaranteed to have 35 bits -- is this problematic?

[the remark this answers is not in the document]
Comment (sipa) [reply src/net_processing.cpp:1708]: Uh, right. I think that could be a problem. This will need something like

```c++
uint64_t time_addr = (uint64_t{current_time / std::chrono::seconds{1}} + hashAddr) / (ROTATE_ADDR_RELAY_DEST_INTERVAL / std::chrono::seconds{1});
```

(I prefer using `x / std::chrono::seconds{1}` over `x.count()` to get the number of seconds of `x`, as the former expression remains meaningful if the type of `x` changes to something other than `std::chrono::seconds`).

File: src/net_processing.cpp [hunk 11]
```diff
@@ -1700,7 +1702,11 @@ void PeerManagerImpl::RelayAddress(NodeId originator,
...
+    const CSipHasher hasher = m_connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write(time_addr);
```
Comment (sipa) [inline src/net_processing.cpp:1709]: No need for the `<< 32` here, I think.

File: src/net_processing.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net_processing.cpp:1708]: My understanding was that `std::chrono::seconds{int64_t{}}` wouldn't compile on platforms where seconds isn't backed by at least 64 bits. I don't expect any such platform to exist.

@ sipa there is also a `count_seconds()` helper.

**Comment by MarcoFalke** [comment]
Just to note `GetTime` is already mockable, so this only changes the type. Though, if this is too much hassle to get right and too hard to review (due to static_cast and `{}`-cast hell), then maybe leave the code as is and only add the test?

File: src/net_processing.cpp (hunk 11 above)
Comment (MarcoFalke) [reply src/net_processing.cpp:1709]: I've pulled that out into #23970, since it also resolves a sanitizer issue.

Comment (naumenkogs) [reply src/net_processing.cpp:1709]: @MarcoFalke does it mean i can leave it as it is here?

File: src/net_processing.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (naumenkogs) [reply src/net_processing.cpp:1708]: Github is weird, I found this un-resolved issue totally accidentally...

File: src/net_processing.cpp (hunk 11 above)
Comment (MarcoFalke) [reply src/net_processing.cpp:1709]: You can either leave it as is or take it, but if you take it, make sure to also remove the suppression like in #23970.

File: src/net_processing.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (naumenkogs) [reply src/net_processing.cpp]: Okay the latest version should be safe?

File: src/net_processing.cpp [hunk 12]
```diff
@@ -124,7 +124,9 @@ static const unsigned int NODE_NETWORK_LIMITED_MIN_BLOCKS = 288;
 /** Average delay between local address broadcasts */
 static constexpr auto AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL = 24h;
 /** Average delay between peer address broadcasts */
-static constexpr auto AVG_ADDRESS_BROADCAST_INTERVAL = 30s;
+static constexpr std::chrono::seconds AVG_ADDRESS_BROADCAST_INTERVAL{30};
```
Comment (MarcoFalke) [inline src/net_processing.cpp:127]: why is this changed? This also conflicts with #23832

Comment (naumenkogs) [reply src/net_processing.cpp:127]: I think it was like that from the very beginning? But sure, I can drop this.

Comment (sipa) [reply src/net_processing.cpp:127]: Using the `static constexpr auto NAME = <numer><suffix>;` as is used here is a bit more readable I think. I'd suggest using the same for the line below.

File: src/net_processing.cpp [hunk 13]
```diff
@@ -1746,8 +1748,12 @@ void PeerManagerImpl::RelayAddress(NodeId originator,
     // Relay to a limited number of other nodes
     // Use deterministic randomness to send to the same nodes for 24 hours
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
-    const uint64_t hashAddr{addr.GetHash()};
-    const CSipHasher hasher{m_connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr).Write((GetTime() + hashAddr) / (24 * 60 * 60))};
+    uint64_t hashAddr = addr.GetHash();
```
Comment (MarcoFalke) [inline src/net_processing.cpp:1751]: why is this changed?

**Review by jonatack — COMMENTED** [review]
@naumenkogs if helpful, here is a rebase of this pull: https://github.com/jonatack/bitcoin/commits/pr18642-rebase-to-master-at-e04720e. If you don't have time to update this in a week or so, I'll open a PR for you with the same commits.

File: src/net_processing.cpp (hunk 13 above)
Comment (jonatack) [reply src/net_processing.cpp]: left that line unchanged in the rebase at https://github.com/jonatack/bitcoin/commit/77ccb7fce15e340080f14c7626cf3dc63fcdee88

**Review by jonatack — COMMENTED** [review]
ACK 2ff8f4dd81dc484fe38ddd9db63cc8fd30192245
